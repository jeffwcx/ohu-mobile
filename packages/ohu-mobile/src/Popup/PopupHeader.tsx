import Button from '../Button';
import { CloseOutlined } from '@ohu-mobile/icons';
import { defineComponent, props } from '../_utils/defineComponent';
import { PopupHeaderProps, PopupHeaderEvents } from './types';
import localeMixin from '../_utils/localeMixin';
import { IconDef } from '../types';

export default defineComponent<PopupHeaderProps, PopupHeaderEvents>(
  'popup-header',
)
  .mixin(localeMixin('OhuPopupHeader'))
  .create({
    props: {
      title: String,
      minorText: String,
      center: props(Boolean).default(false),
      confirm: props(Boolean).default(false),
      closeIcon: props.ofType<IconDef>().default(() => CloseOutlined),
      closeIconPosition: props
        .ofStringLiterals('left', 'right')
        .default('right'),
    },
    methods: {
      close() {
        const parent = this.$parent as any;
        if (parent.close) {
          parent.close();
        }
        this.$emit('cancel');
      },
      handleConfirm() {
        this.$emit('confirm');
      },
    },
    render() {
      const root = this.$rootCls();
      const textNode = root.element('text');
      const {
        $slots,
        title,
        minorText,
        center,
        confirm,
        closeIcon,
        closeIconPosition,
      } = this;
      const content = (
        <div class={textNode}>
          <div class={textNode.element('title').has([!!minorText && 'minor'])}>
            {$slots.default || title}
          </div>
          {minorText && (
            <div class={textNode.element('minor-title')}>{minorText}</div>
          )}
        </div>
      );
      if (confirm) {
        return (
          <div class={root.is(['confirm', center && 'text-center'])}>
            <Button
              class={root.element('cancel')}
              link
              inline
              size="sm"
              onClick={this.close}
            >
              {this.$l.cancelText}
            </Button>
            {content}
            <Button
              class={root.element('ok')}
              type="primary"
              link
              inline
              size="sm"
              onClick={this.handleConfirm}
            >
              {this.$l.confirmText}
            </Button>
          </div>
        );
      }
      return (
        <div
          class={root.is([
            center && 'text-center',
            `close-icon-${closeIconPosition}`,
          ])}
        >
          {content}
          <Button
            type="link"
            class={root.element('close')}
            size="md"
            inline
            icon={closeIcon}
            onClick={this.close}
          ></Button>
        </div>
      );
    },
  });
