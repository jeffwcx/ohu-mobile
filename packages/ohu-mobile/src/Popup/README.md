# Popup
## Usage

```tsx
import { Popup, component } from 'ohu-mobile';

component({
  data() {
    return {
      visible: false,
    };
  },
  render() {
    return (
      <Popup v-model={this.visible}></Popup>
    );
  },
})
```

## Function Call

```ts
Popup.open();
```



# TODO

+ dymanic width and height in anchor mode.
+ round radius + gesture suport.
+ bounce build animation.


