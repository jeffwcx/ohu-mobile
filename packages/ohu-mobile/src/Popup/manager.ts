export default {
  zIndex: 9999,
  isLock: 0,
  getPopupZIndex() {
    const maskZIndex = this.zIndex + 1;
    const documentZIndex = this.zIndex + 2;
    this.zIndex += 2;
    return {
      maskZIndex,
      documentZIndex,
    };
  },
};
