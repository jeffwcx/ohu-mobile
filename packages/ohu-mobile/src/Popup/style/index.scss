@use '../../_styles/variables' as *;
@use '../../_styles/mixins' as *;
$maskPrefixClass: #{$prefix}mask;

.#{$maskPrefixClass} {
  position: fixed;
  width: 100%;
  height: 100%;
  background: $colorMask;
  touch-action: none;
  -webkit-tap-highlight-color: transparent;
  // need compatible
  &.is-frosted {
    backdrop-filter: blur(20Px);
  }
}


$popupPrefixClass: #{$prefix}popup;
.#{$popupPrefixClass} {
  -webkit-tap-highlight-color: transparent;
  &.is-fullscreen {
    width: 100%;
    height: 100%;
  }
  &.is-round {
    border-radius: $popupRoundRadius $popupRoundRadius 0 0;
    overflow: hidden;
  }
  &-wrapper {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    &.is-tap-through {
      pointer-events: none;
    }
    &.is-scrollable {
      overscroll-behavior-y: contain;
      overflow: auto;
      text-align: center;
      align-items: flex-start;
      &::after {
        content: '';
        width: 0;
        height: 100%;
        display: inline-block;
        vertical-align: middle;
      }
      .#{$prefix}popup {
        display: inline-block;
        vertical-align: middle;
        text-align: left;
      }
    }
    &.has-position {
      display: flex;
      flex-flow: row nowrap;
    }
    &.is-x-center {
      justify-content: center;
    }
    &.is-y-center {
      align-items: center;
    }
    &.is-x-left {
      justify-content: flex-start;
    }
    &.is-x-right {
      justify-content: flex-end;
    }
    &.is-y-top {
      align-items: flex-start;
    }
    &.is-y-bottom {
      align-items: flex-end;
    }
  }
}

@import './popup-header.scss';
