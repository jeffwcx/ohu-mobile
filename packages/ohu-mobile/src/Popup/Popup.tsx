import { getScrollEventTarget } from '../_utils/dom';
import manager from './manager';
import debounce from '../_utils/debounce';
import { PopupTransformOrigin, PopupPosition,
  PopupAnimateType, PopupEvents,
  PopupAnchorPosition, PopupEnterEvent, PopupProps,
} from './types';
import { isAnyPosition, computeRect, getAnchorPosition, getTransformOrigin } from './utils';
import { $prefix } from '../_config/variables';
import { defineComponent, props } from '../_utils/defineComponent';
import { ClassOptions } from '../_utils/classHelper';
import { CSSProps } from '../types';
import bindEvent from '../_utils/bindEvent';


interface PopupData {
  wrapperVisible: boolean;
  documentVisible: boolean;
  docEl: HTMLElement | null;
  docRect: { width: number, height: number } | null;
  anchorEl: HTMLElement | null;
  docPos: {
    top: number,
    bottom: number,
    left: number,
    right: number,
    transformOrigin: string,
  } | null;
  anchorRect: DOMRect | ClientRect | null;
  maskZIndex: number;
  documentZIndex: number;
  touchData: {
    scrollY: number,
    pageY: number,
    scrollEl: HTMLElement | null,
    maxHeight: number,
  };
  rootStyle: CSSProps;
}

export const popupProps = {
  visible: Boolean,
  anchor: props.ofType<HTMLElement | (() => HTMLElement)>().optional,
  transformOrigin: props.ofType<PopupTransformOrigin>().optional,
  marginThreshold: props(Number).default(12),
  edgeDetect: props(Boolean).default(true),
  lockScroll: props(Boolean).default(true),
  position: props.ofType<PopupPosition>().default('center'),
  mask: props(Boolean).default(true),
  maskFrosted: props(Boolean).default(false),
  maskClosable: props(Boolean).default(true),
  maskAnimate: props.ofStringLiterals('mask-fade', 'none').default('mask-fade'),
  partialMask: props.ofStringLiterals('top', 'bottom').optional,
  closeOnMaskTouched: props(Boolean).default(false),
  fullscreen: props(Boolean).default(false),
  animate: props.ofType<PopupAnimateType>().default('none'),
  targetStyle: props.ofType<CSSProps>().optional,
  targetClass: props.ofType<ClassOptions>().optional,
  scrollBody: props(Boolean).default(false),
  tapThrough: props(Boolean).default(false),
  zIndex: Number,
  round: props(Boolean).default(false),
  hideOnDeactivated: props(Boolean).default(true),
}

const positionTransitionMap = {
  top: `${$prefix}slide-up`,
  bottom: `${$prefix}slide-down`,
  left: `${$prefix}slide-left`,
  right: `${$prefix}slide-right`,
};

const createPopup = defineComponent<PopupProps, PopupEvents>('popup-main');

export const POPUP_EVENT = 'visibleChange';
const Popup = createPopup.create({
  model: {
    prop: 'visible',
    event: POPUP_EVENT,
  },
  props: popupProps,
  data() {
    return {
      wrapperVisible: this.visible,
      documentVisible: false,
      docRect: null,
      docEl: null,
      docPos: null,
      anchorEl: null,
      anchorRect: null,
      maskZIndex: 0,
      documentZIndex: 0,
      touchData: {
        scrollY: 0,
        pageY: 0,
        scrollEl: null,
        maxHeight: 0,
      },
      rootStyle: {},
    } as PopupData;
  },
  watch: {
    visible(newValue) {
      newValue ? this.open() : this.close();
    },
  },
  computed: {
    isAnchorMode() {
      return !!this.anchor;
    },
    documentClass() {
      return this.bem.block('popup')
        .is([
          this.fullscreen && 'fullscreen',
          this.round && 'round',
        ])
        .addClasses(this.targetClass);
    },
    documentTransition() {
      if (this.animate === 'none'
        && typeof this.position === 'string'
        && this.position !== 'center') {
        return positionTransitionMap[this.position];
      }
      return `${$prefix}${this.animate}`;
    },
  },
  created() {
    this.$on('hook:activated', () => {
      this.rootStyle = {};
      if (this.hideOnDeactivated && manager.isLock > 0) {
        document.body.style.overflow = 'hidden';
      }
    });
    this.$on('hook:deactivated', () => {
      if (this.hideOnDeactivated) {
        this.rootStyle = { display: 'none' };
        if (manager.isLock > 0) {
          document.body.style.overflow = '';
        }
      }
    });
    const resizeHandler = debounce(() => {
      if (this.documentVisible) {
        this.computeDocAndAnchorRect();
      }
    });
    this.$on('hook:mounted', () => {
      this.visible ? this.open() : this.close();
      bindEvent(this, 'resize', resizeHandler);
      bindEvent(this, 'scroll', resizeHandler);
    });
    this.$on('hook:beforeDestroy', () => {
      if (this.documentVisible) {
        this.close();
      }
    });
  },
  methods: {
    // ⬇️ open api
    close() {
      if (this.documentVisible) {
        this.documentVisible = false;
        this.$emit(POPUP_EVENT, this.documentVisible);
        this.$emit('close', this.documentVisible);
        this.startScroll();
      }
    },
    open() {
      if (this.documentVisible) return;
      let maskZIndex;
      let documentZIndex;
      if (this.zIndex !== undefined) {
        maskZIndex = this.zIndex;
        documentZIndex = this.zIndex + 1;
      } else {
        const result = manager.getPopupZIndex();
        maskZIndex = result.maskZIndex;
        documentZIndex = result.documentZIndex;
      }
      this.maskZIndex = maskZIndex;
      this.documentZIndex = documentZIndex;
      this.documentVisible = true;
      this.stopScroll();
      this.$emit('open', {
        visible: this.documentVisible,
        documentZIndex: this.documentZIndex,
        maskZIndex: this.maskZIndex,
      });
    },
    // ⬆️ open api
    stopScroll() {
      if (this.lockScroll) {
        manager.isLock += 1;
        document.body.style.overflow = 'hidden';
      }
    },
    startScroll() {
      if (manager.isLock === 0) return;
      if (this.lockScroll) {
        manager.isLock -= 1;
        if (manager.isLock <= 0) {
          document.body.style.overflow = '';
        }
      }
    },
    onMaskClick(e: Event) {
      e.stopPropagation();
      if (this.maskClosable) {
        this.close();
      }
    },
    onWrapperTouchstart(e: TouchEvent) {
      if (this.closeOnMaskTouched) {
        this.close();
        return;
      }
      const event = e.touches[0];
      const el = e.target as HTMLElement;
      const rootEl = this.$refs.document as HTMLElement;
      if (rootEl) {
        const scrollEl = getScrollEventTarget(el, rootEl) as HTMLElement;
        this.touchData.scrollEl = scrollEl;
        this.touchData.pageY = event.pageY;
        this.touchData.scrollY = scrollEl.scrollTop;
        this.touchData.maxHeight = scrollEl.scrollHeight - scrollEl.clientHeight;
      }
    },
    onWrapperTouchmove(e: TouchEvent) {
      if (this.scrollBody) return;
      const scrollEl = this.touchData.scrollEl;
      const event = e.touches[0];
      if (scrollEl) {
        const { pageY, maxHeight } = this.touchData;
        const scrollTop = scrollEl.scrollTop;
        const distanceY = event.pageY - pageY;
        if (distanceY > 0 && scrollTop <= 0) {
          e.cancelable && e.preventDefault();
          return;
        }
        if (distanceY < 0 && scrollTop + 1 >= maxHeight) {
          e.cancelable && e.preventDefault();
          return;
        }
      }
    },
    onWrapperTouchend() {
      this.touchData.maxHeight = 0;
      this.touchData.pageY = 0;
      this.touchData.scrollEl = null;
      this.touchData.scrollY = 0;
    },
    onAfterDocumentLeave() {
      this.wrapperVisible = false;
      this.$emit('afterLeave', this.wrapperVisible);
    },
    onBeforeDocumentEnter() {
      this.wrapperVisible = true;
    },
    onDocumentClick(e: Event) {
      e.stopPropagation();
    },
    onDocumentTouch(e: TouchEvent) {
      if (this.closeOnMaskTouched) {
        e.stopPropagation();
      }
    },
    onDocumentEnter(doc: Element) {
      this.docEl = doc as HTMLElement;
      this.$nextTick(() => {
        this.computeDocAndAnchorRect();
        const enterEvent: PopupEnterEvent = {};
        if (this.docRect) {
          enterEvent.doc = Object.assign({}, this.docPos, this.docRect);
        }
        if (this.anchorRect) {
          enterEvent.anchor = this.anchorRect;
        }
        this.$emit('enter', enterEvent);
      });
    },
    onDocumentAfterEnter() {
      this.$emit('afterOpen');
    },
    computeDocAndAnchorRect() {
      if (this.docEl) {
        // use offsetWidth and offsetHeight, when getBoudingClientRect is not correct.
        this.docRect = computeRect(this.docEl);
      }
      if (this.anchor) {
        if (this.anchor instanceof Function) {
          this.anchorEl = this.anchor();
        } else {
          this.anchorEl = this.anchor;
        }
        if (this.anchorEl) {
          this.anchorRect = this.anchorEl.getBoundingClientRect();
        }
      }
      this.docPos = this.computeDocumentPosition();
    },
    renderMask() {
      const maskCls = this.bem.block('mask')
        .is(this.maskFrosted && 'frosted');
      const maskStyle: Partial<CSSStyleDeclaration> = {
        zIndex: this.maskZIndex.toString(),
        top: '0px',
        left: '0px',
      };
      if (this.docPos) {
        if (this.partialMask === 'bottom') {
          maskStyle.height = 'auto';
          maskStyle.top = `${this.docPos.top}px`;
          maskStyle.bottom = '0';
        } else if (this.partialMask === 'top') {
          maskStyle.height = 'auto';
          maskStyle.bottom = `${this.docPos.bottom}px`;
          maskStyle.top = '0';
        }
      }
      let maskNode = <div v-show={this.documentVisible} class={maskCls} style={maskStyle}></div>;
      if (this.maskAnimate !== 'none') {
        maskNode = (
          <transition name={$prefix + this.maskAnimate}>
            {maskNode}
          </transition>
        );
      }
      if (this.mask) {
        return maskNode;
      }
    },
    computeDocumentPosition() {
      const { position, isAnchorMode } = this;
      let top = 0;
      let left = 0;
      let transformOriginStr = '';
      if (isAnchorMode) {
        const { anchorRect, docRect } = this;
        if (anchorRect) {
          // transform anchor position
          let anchorPos: PopupAnchorPosition = getAnchorPosition(position);
          let anchorTop = anchorRect.top;
          let anchorLeft = anchorRect.left;
          if (anchorPos.horizontal === 'center') {
            anchorLeft = anchorLeft + anchorRect.width / 2;
          } else if (anchorPos.horizontal === 'right') {
            anchorLeft = anchorLeft + anchorRect.width;
          }
          if (anchorPos.vertical === 'center') {
            anchorTop = anchorTop + anchorRect.height / 2;
          } else if (anchorPos.vertical === 'bottom') {
            anchorTop = anchorTop + anchorRect.height;
          }
          const transformOrigin = getTransformOrigin(anchorPos, this.transformOrigin);
          transformOriginStr = (transformOrigin.vertical || 'top') + ' ' + (transformOrigin.horizontal || 'left');
          let docOffsetX = 0;
          let docOffsetY = 0
          if (docRect) {
            if (transformOrigin.horizontal === 'right') {
              docOffsetX = docRect.width;
            } else if (transformOrigin.horizontal === 'center') {
              docOffsetX = docRect.width / 2;
            }
            if (transformOrigin.vertical === 'bottom') {
              docOffsetY = docRect.height;
            } else if (transformOrigin.vertical === 'center') {
              docOffsetY = docRect.height / 2;
            }
          }
          top = anchorTop - docOffsetY;
          left = anchorLeft - docOffsetX;
        }
      } else if(isAnyPosition(position)) {
        position.left !== undefined && (left = position.left);
        position.top !== undefined && (top = position.top);
      }
      let lowerEdge = window.innerHeight - top;
      let rightEdge = window.innerWidth - left;
      if (this.docRect) {
        lowerEdge = lowerEdge - this.docRect.height;
        rightEdge = rightEdge - this.docRect.width;
      }

      // container edge detect
      if (this.edgeDetect) {
        const widthThreshold = window.innerWidth - this.marginThreshold;
        const heightThreshold = window.innerHeight - this.marginThreshold;
        if (this.docRect) {
          let bottom = top + this.docRect.height;
          let right = left + this.docRect.width;
          if (top < this.marginThreshold) {
            top = this.marginThreshold;
          } else if (bottom > heightThreshold) {
            top = heightThreshold - this.docRect.height;
          }
          if (left < this.marginThreshold) {
            left = this.marginThreshold;
          } else if (right > widthThreshold) {
            left = widthThreshold - this.docRect.width;
          }
        }
      }
      return {
        top,
        left,
        bottom: lowerEdge,
        right: rightEdge,
        transformOrigin: transformOriginStr,
      };
    },
    getDocumentStyle() {
      let docStyle: Partial<CSSStyleDeclaration> = {};
      if (this.isAnchorMode) {
        docStyle = {
          position: 'absolute',
        };
        if (this.docPos) {
          docStyle.top = `${this.docPos.top}px`;
          docStyle.left = `${this.docPos.left}px`;
          docStyle.transformOrigin = this.docPos.transformOrigin;
        }
      }
      if (this.targetStyle) {
        docStyle = {
          ...docStyle,
          ...this.targetStyle,
        };
      }
      return docStyle;
    },

    renderDocument() {
      const wrapperCls = this.bem.block('popup-wrapper');
      wrapperCls.is([
        this.tapThrough && 'tap-through',
        this.scrollBody && 'scrollable',
      ]);
      if (!this.isAnchorMode) {
        const { horizontal, vertical } = getAnchorPosition(this.position);
        if (horizontal || vertical) {
          wrapperCls.has('position');
        }
        if (horizontal) {
          wrapperCls.is(`x-${horizontal}`);
        }
        if (vertical && !this.scrollBody) {
          wrapperCls.is(`y-${vertical}`);
        }
      }
      const wrapperStyle = {
        zIndex: this.documentZIndex,
      };
      return (
        <div v-show={this.wrapperVisible}
          class={wrapperCls}
          style={wrapperStyle}
          role="dialog"
          onClick={this.onMaskClick}
          onTouchstart={this.onWrapperTouchstart}
          onTouchmove={this.onWrapperTouchmove}
          onTouchend={this.onWrapperTouchend}
          tabindex={-1}>
          <transition
            name={this.documentTransition}
            onBeforeEnter={this.onBeforeDocumentEnter}
            onEnter={this.onDocumentEnter}
            onAfterEnter={this.onDocumentAfterEnter}
            onAfterLeave={this.onAfterDocumentLeave}>
            <div v-show={this.documentVisible}
              role="document"
              ref="document"
              class={this.documentClass}
              style={this.getDocumentStyle()}
              onTouchstart={this.onDocumentTouch}
              onTouchmove={this.onDocumentTouch}
              onTouchend={this.onDocumentTouch}
              onClick={this.onDocumentClick}>
              { this.$slots.default }
            </div>
          </transition>
        </div>
      );
    },
  },
  render() {
    return (
      <div role="presentation" style={this.rootStyle}>
        { this.renderMask() }
        { this.renderDocument() }
      </div>
    );
  }
});

export default Popup;
