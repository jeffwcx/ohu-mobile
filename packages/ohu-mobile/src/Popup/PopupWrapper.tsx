import PortalRender from '../_utils/PortalRender';
import Popup, { popupProps, POPUP_EVENT } from './Popup';
import { VNodeData } from 'vue';
import { PopupOutSideEvents, PopupOutSideProps, PopupGetContainerFunc } from './types';
import { defineComponent, props } from '../_utils/defineComponent';

export const popupOutSideProps = {
  getContainer: props.ofType<PopupGetContainerFunc>().default(() => document.body),
  dynamic: props(Boolean).default(false),
  usePortal: props(Boolean).default(true),
  ...popupProps,
};

const createPopupWrapper = defineComponent<PopupOutSideProps, PopupOutSideEvents>('popup');

const PopupWrapper = createPopupWrapper.create({
  model: {
    prop: 'visible',
    event: POPUP_EVENT,
  },
  props: popupOutSideProps,
  methods: {
    close() {
      (this.$refs.popup as any).close();
    },
    open() {
      (this.$refs.popup as any).open();
    },
  },
  render() {
    const {
      $props,
      $listeners,
      $attrs,
      $slots,
      visible,
      getContainer,
      dynamic,
    } = this;
    const popupNodeData: VNodeData = {
      props: $props,
      on: {
        ...$listeners,
        afterLeave: () => {
          if (dynamic && this.$refs.portal) {
            (this.$refs.portal as any).remove();
          }
          this.$emit('afterClose');
        },
      },
      attrs: $attrs,
      ref: 'popup',
    };
    if (!this.usePortal) return (
      <Popup {...popupNodeData}>{$slots.default}</Popup>
    );
    const container = getContainer instanceof Function
      ? getContainer()
      : getContainer;
    return (
      <PortalRender
        ref="portal"
        visible={visible}
        container={container}
        children={() => (
          <Popup {...popupNodeData}>{$slots.default}</Popup>
        )}>
      </PortalRender>
    )
  },
});

export default PopupWrapper;
