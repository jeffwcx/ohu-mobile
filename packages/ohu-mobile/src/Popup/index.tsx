import PopupWrapper from './PopupWrapper';
import Vue from 'vue';
import { VueConstructor } from 'vue/types/vue';
import { PopupOpenOptions } from './types';
import PopupHeader from './PopupHeader';
import manager from './manager';
import noop from '../_utils/noop';

let instance: InstanceType<VueConstructor> & { visible: boolean };
const Popup = Object.assign(PopupWrapper, {
  Header: PopupHeader,
  open: (props: PopupOpenOptions) => {
    const {
      parent,
      render,
      onEnter,
      onOpen,
      onAfterClose,
      onVisibleChange,
      onClose,
      ...popupProps
    } = props;
    instance = new Vue({
      parent: parent || undefined,
      el: document.createElement('div'),
      data() {
        return {
          visible: true,
        };
      },
      render(h) {
        const popupData = {
          props: {
            ...popupProps,
            visible: this.visible,
            dynamic: true,
            usePortal: true,
          },
          on: {
            enter: onEnter || noop,
            open: onOpen || noop,
            visibleChange: onVisibleChange || noop,
            close: onClose || noop,
            afterClose: (e: any) => {
              if (onAfterClose) onAfterClose(e);
              this.$destroy();
            },
          },
        };
        return <Popup {...popupData}>{render && render(h)}</Popup>;
      },
    });
  },
  close: () => {
    if (instance) {
      instance.visible = false;
    }
  },
});

export default Popup;

export * from './types';

export { PopupHeader, manager };
