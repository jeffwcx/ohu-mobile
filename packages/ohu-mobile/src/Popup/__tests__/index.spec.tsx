import { CloseOutlined } from '@ohu-mobile/icons';
import { config, mount } from '@vue/test-utils';
import Vue from 'vue';
import Popup from '..';
import Button from '../../Button';
import { wait } from '../../_utils/test';
import VueRouter from 'vue-router';

config.stubs!.transition = false;


function resizeWindow(width: number, height: number) {
  Object.defineProperties(window, {
    innerWidth: { value: width, configurable: true },
    innerHeight: { value: height, configurable: true },
  });
  window.dispatchEvent(new Event('resize'));
}

function createDiv() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  return div;
}

describe('Popup', () => {
  it('v-model and popup-header', async () => {
    const div = createDiv();
    const onClose = jest.fn();
    const onConfirm = jest.fn();
    const wrapper = mount(Vue.extend({
      data() {
        return {
          visible: false,
        };
      },
      render() {
        return (
          <Popup
            v-model={this.visible}
            usePortal={false}
            maskClosable
            zIndex={1000}
            onClose={onClose}
            targetStyle={{ width: '100px', height: '100px' }}>
            <Popup.Header
              title="标题"
              minorText="minorText"
              center
              closeIcon={CloseOutlined}
              confirm
              onConfirm={onConfirm}>
            </Popup.Header>
          </Popup>
        );
      },
    }), {
      attachTo: div,
    });
    wrapper.setData({ visible: true });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    await wrapper.find('.ohu-popup-header__cancel').trigger('click');
    expect(onClose).toBeCalledTimes(1);
    expect(wrapper.vm.visible).toBeFalsy();
    await wrapper.find('.ohu-popup-header__ok').trigger('click');
    expect(onConfirm).toBeCalledTimes(1);
    wrapper.destroy();
  });

  it('should add class and style to wrapper when using `targetClass` and `targetStyle`', async () => {
    const div = createDiv();
    const onOpen = jest.fn();
    const wrapper = mount(Popup, {
      propsData: {
        visible: true,
        targetClass: 'cls',
        targetStyle: { width: '50%', height: '200px' },
        usePortal:false,
        zIndex: 1001,
      },
      listeners: {
        open: onOpen,
      },
      slots: {
        default: 'text',
      },
      attachTo: div,
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(onOpen).toBeCalledTimes(1);
  });

  it('should detect edge in anchor mode', async () => {
    const div = createDiv();
    const mockAnchor = {
      getBoundingClientRect: jest.fn().mockReturnValue({
        bottom: 391,
        height: 48,
        left: 22,
        right: 101,
        top: 343,
        width: 80,
        x: 22,
        y: 343,
      }),
    };
    const wrapper = mount(Vue.extend({
      data() {
        return {
          visible: false,
        };
      },
      render() {
        return (
          <div>
            <Popup
              v-model={this.visible}
              mask
              maskClosable
              position="right"
              transformOrigin={{ vertical: 'bottom', horizontal: 'center' }}
              anchor={() => mockAnchor as any}
              usePortal={false}
              edgeDetect>
              <p>anchor and edge detect</p>
            </Popup>
            <Button onClick={() => this.visible = true}>anchor</Button>
          </div>
        );
      },
    }), {
      attachTo: div,
    });
    await wrapper.find('.ohu-btn').trigger('click');
    await Vue.nextTick();
    const popup = wrapper.find('.ohu-popup').element
    expect(popup.style['top']).toBe('367px');
    expect(popup.style['left']).toBe('102px');
    expect(popup.style['transformOrigin']).toBe('bottom center');
    resizeWindow(100, 500);
    await wait(600);
    expect(popup.style['left']).toBe('88px');
  });

  it('should emit `close` event when mask is clicked', async () => {
    const div = createDiv();
    const onClose = jest.fn();
    const wrapper = mount(Popup, {
      propsData: {
        visible: true,
        zIndex: 1002,
        position: 'bottom',
        fullscreen: true,
        maskFrosted: true,
        usePortal: false,
        tapThrough: true,
      },
      listeners: {
        close: onClose,
      },
      slots: {
        default: Vue.extend({
          render() {
            return (
              <div style="width: 100px;height: 200px;">mask</div>
            );
          },
        }),
      },
      attachTo: div,
    });

    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    await wrapper.find('[role="dialog"]').trigger('click');
    expect(onClose).toBeCalledTimes(1);
  });


  it('should add component to body when use call `open()`', async () => {
    const div = createDiv();
    const afterClose = jest.fn();
    jest.spyOn(document.body, 'appendChild');
    const wrapper = mount(Vue.extend({
      methods: {
        open() {
          Popup.open({
            parent: this,
            animate: 'none',
            onAfterClose: afterClose,
            render() {
              return <p>Function Call</p>;
            },
          });
        },
        close() {
          Popup.close();
        }
      },
      render() {
        return (
          <div>
          </div>
        );
      },
    }), {
      attachTo: div,
    });
    await Vue.nextTick();
    wrapper.vm.open();
    await wait(300);
    expect(document.body.appendChild).toBeCalledWith(expect.objectContaining({
      title: 'TEST',
    }));
    wrapper.vm.close();
    await wait(300);
    expect(afterClose).toBeCalled();
  });

  it('should be closed when mask has been touched', async () => {
    const div = createDiv();
    const onClose = jest.fn();
    const wrapper = mount(Popup, {
      propsData: {
        visible: true,
        closeOnMaskTouched: true,
        fullscreen: true,
        usePortal: false,
        partialMask: 'top',
      },
      slots: {
        default: 'text',
      },
      listeners: {
        close: onClose,
      },
      attachTo: div,
    });
    await Vue.nextTick();
    wrapper.find('[role="document"]').trigger('touchstart');
    wrapper.find('[role="dialog"]').trigger('touchstart');
    await wrapper.find('[role="dialog"]').trigger('touchend');
    expect(onClose).toBeCalled();
  });


  // https://github.com/jeffwcx/ohu-mobile/issues/23
  it('should hide when it changes from deactivated to activated', async () => {
    const div = createDiv();
    const A = Vue.extend({
      render() {
        return (
          <div>
            <Popup
              visible
              scrollBody
              usePortal={false}
              targetStyle={{ height: '120vh', width: '100px' }}>
              A
            </Popup>
          </div>
        );
      },
    });
    const B = Vue.extend({
      render() {
        return <div>B</div>
      },
    });
    const router = new VueRouter({
      mode: 'hash',
      routes: [
        { path: '/', name: 'a', component: A },
        { path: '/b', name: 'b', component: B },
      ],
    });
    Vue.use(VueRouter);
    const wrapper = mount(Vue.extend({
      router,
      methods: {
        pushToB() {
          router.push('/b');
        },
      },
      render() {
        return (
          <keep-alive>
            <router-view></router-view>
          </keep-alive>
        );
      },
    }), {
      attachTo: div,
    });
    await wait(300);
    expect(document.body.style.overflow).toBe('hidden');
    wrapper.vm.pushToB();
    await wait(300);
    expect(document.body.style.overflow).toBe('');
  });

});
