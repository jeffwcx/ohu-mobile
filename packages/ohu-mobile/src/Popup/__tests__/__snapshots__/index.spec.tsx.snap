// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popup > should add class and style to wrapper when using \`targetClass\` and \`targetStyle\` 1`] = `
"<div role="presentation">
  <div class="ohu-mask ohu-mask-fade-enter ohu-mask-fade-enter-active" style="z-index: 1001; top: 0px; left: 0px;"></div>
  <div role="dialog" tabindex="-1" class="ohu-popup-wrapper has-position is-x-center is-y-center" style="z-index: 1002; z-index: 1002;">
    <div role="document" class="ohu-popup cls ohu-none-enter ohu-none-enter-active" style="width: 50%; height: 200px;">text</div>
  </div>
</div>"
`;

exports[`Popup > should emit \`close\` event when mask is clicked 1`] = `
"<div role="presentation">
  <div class="ohu-mask is-frosted ohu-mask-fade-enter ohu-mask-fade-enter-active" style="z-index: 1002; top: 0px; left: 0px;"></div>
  <div role="dialog" tabindex="-1" class="ohu-popup-wrapper is-tap-through has-position is-x-center is-y-bottom" style="z-index: 1003; z-index: 1003;">
    <div role="document" class="ohu-popup is-fullscreen ohu-slide-down-enter ohu-slide-down-enter-active" style="">
      <div style="width: 100px; height: 200px;">mask</div>
    </div>
  </div>
</div>"
`;

exports[`Popup > v-model and popup-header 1`] = `
"<div role="presentation">
  <div class="ohu-mask ohu-mask-fade-enter ohu-mask-fade-enter-active" style="z-index: 1000; top: 0px; left: 0px;"></div>
  <div role="dialog" tabindex="-1" class="ohu-popup-wrapper has-position is-x-center is-y-center" style="z-index: 1001; z-index: 1001;">
    <div role="document" class="ohu-popup ohu-none-enter ohu-none-enter-active" style="width: 100px; height: 100px;">
      <div class="ohu-popup-header is-confirm is-text-center"><button role="button" tabindex="0" type="button" class="ohu-btn is-default is-link is-sm is-inline ohu-popup-header__cancel"></button>
        <div class="ohu-popup-header__text">
          <div class="ohu-popup-header__text__title has-minor">标题</div>
          <div class="ohu-popup-header__text__minor-title">minorText</div>
        </div><button role="button" tabindex="0" type="button" class="ohu-btn is-primary is-link is-sm is-inline ohu-popup-header__ok"></button>
      </div>
    </div>
  </div>
</div>"
`;
